client/widgets/markets/TradePanels.cpp

```diff
@@ -0,0 +1,394 @@
+#include "../../CGameInfo.h"
+#include "../../CPlayerInterface.h"
+
+#include "../../../CCallback.h"
+
+#include "../../../lib/CGeneralTextHandler.h"
+#include "../../../lib/mapObjects/CGHeroInstance.h"
+
+CTradeableItem::CTradeableItem(Point pos, EType Type, int ID, bool Left, int Serial)
+	: CIntObject(LCLICK | HOVER | SHOW_POPUP, pos)
+	, hlp(nullptr)
+	, type(EType(-1)) // set to invalid, will be corrected in setType
+	, id(ID)
+	, serial(Serial)
+	, left(Left)
+	, downSelection(false)
+{
+	OBJECT_CONSTRUCTION_CAPTURING(255 - DISPOSE);
+	
+	setType(Type);
+	if(image)
+	{
+		this->pos.w = image->pos.w;
+		this->pos.h = image->pos.h;
+	}
+}
+
+void CTradeableItem::setType(EType newType)
+{
+	if(type != newType)
+	{
+		OBJECT_CONSTRUCTION_CUSTOM_CAPTURING(255 - DISPOSE);
+		type = newType;
+
+		if(getIndex() < 0)
+		{
+			image = std::make_shared<CAnimImage>(getFilename(), 0);
+			image->disable();
+		}
+		else
+		{
+			image = std::make_shared<CAnimImage>(getFilename(), getIndex());
+		}
+	}
+}
+
+void CTradeableItem::setID(int newID)
+{
+	if(id != newID)
+	{
+		id = newID;
+		if(image)
+		{
+			int index = getIndex();
+			if(index < 0)
+				image->disable();
+			else
+			{
+				image->enable();
+				image->setFrame(index);
+			}
+		}
+	}
+}
+
+AnimationPath CTradeableItem::getFilename()
+{
+	switch(type)
+	{
+	case EType::RESOURCE:
+		return AnimationPath::builtin("RESOURCE");
+	case EType::PLAYER:
+		return AnimationPath::builtin("CREST58");
+	case EType::ARTIFACT_TYPE:
+	case EType::ARTIFACT_PLACEHOLDER:
+	case EType::ARTIFACT_INSTANCE:
+		return AnimationPath::builtin("artifact");
+	case EType::CREATURE:
+		return AnimationPath::builtin("TWCRPORT");
+	default:
+		return {};
+	}
+}
+
+int CTradeableItem::getIndex()
+{
+	if(id < 0)
+		return -1;
+
+	switch(type)
+	{
+	case EType::RESOURCE:
+	case EType::PLAYER:
+		return id;
+	case EType::ARTIFACT_TYPE:
+	case EType::ARTIFACT_INSTANCE:
+	case EType::ARTIFACT_PLACEHOLDER:
+		return CGI->artifacts()->getByIndex(id)->getIconIndex();
+	case EType::CREATURE:
+		return CGI->creatures()->getByIndex(id)->getIconIndex();
+	default:
+		return -1;
+	}
+}
+
+void CTradeableItem::showAll(Canvas & to)
+{
+	Point posToBitmap;
+	Point posToSubCenter;
+
+	switch(type)
+	{
+	case EType::RESOURCE:
+		posToBitmap = Point(19, 9);
+		posToSubCenter = Point(35, 57);
+		break;
+	case EType::CREATURE_PLACEHOLDER:
+	case EType::CREATURE:
+		posToSubCenter = Point(29, 77);
+		break;
+	case EType::PLAYER:
+		posToSubCenter = Point(31, 77);
+		break;
+	case EType::ARTIFACT_PLACEHOLDER:
+	case EType::ARTIFACT_INSTANCE:
+		posToSubCenter = Point(22, 51);
+		if (downSelection)
+			posToSubCenter.y += 8;
+		break;
+	case EType::ARTIFACT_TYPE:
+		posToSubCenter = Point(35, 57);
+		posToBitmap = Point(13, 0);
+		break;
+	}
+
+	if(image)
+	{
+		image->moveTo(pos.topLeft() + posToBitmap);
+		CIntObject::showAll(to);
+	}
+
+	to.drawText(pos.topLeft() + posToSubCenter, FONT_SMALL, Colors::WHITE, ETextAlignment::CENTER, subtitle);
```

Comment:
Likely old code, so feel free to ignore / postpone, but:
- might be better to move image once, when type/image is changed, not on every redraw
- drawText should be replaced CLabel

Reply:
> Likely old code

Yes. I my plan is create a separate CLabel for this. And separate image.